### PR TITLE
HUE-9681 [Django Upgrade] desktop.auth.backend.AllowFirstUserDjangoBa…

### PR DESCRIPTION
**Pull request by agl29**
…ckend.authenticate() to accept a positional  argument RemovedInDjango21Warning

### CODE REVIEW

**Review by romainr — COMMENTED** [review]
Now for Django 1.1, and I guess this will break Django 1?

One idea would be to convert to compacted arguments to be backward compatible at the same time, maybe:

```def authenticate(self, *args, **kwargs):
  if py3:
    request = args[0]
   
  username = kwargs['username']
  ...
```

Note: I guess we should also port the other backends similarly one by one too

**Comment by romainr** [comment]
bump

**Comment by agl29** [comment]
As we are using 'request' parameter earlier also like here https://github.com/cloudera/hue/blob/master/desktop/core/src/desktop/auth/backend.py#L404. So we need to change all the `def authentication()` into the above (as you describe) or we can follow what we are doing earlier also (like request=None)

**Comment by romainr** [comment]
> As we are using 'request' parameter earlier also like here https://github.com/cloudera/hue/blob/master/desktop/core/src/desktop/auth/backend.py#L404. So we need to change all the `def authentication()` into the above (as you describe) or we can follow what we are doing earlier also (like request=None)

Not sure option 2 would work, as it need to be a positional argument 1 and in some backend it would need to be after some other args https://github.com/cloudera/hue/blob/master/desktop/core/src/desktop/auth/backend.py#L247

File: desktop/core/src/desktop/auth/backend.py [hunk 1]
```diff
@@ -273,7 +273,14 @@ class AllowFirstUserDjangoBackend(django.contrib.auth.backends.ModelBackend):
   Allows the first user in, but otherwise delegates to Django's
   ModelBackend.
   """
-  def authenticate(self, username=None, email=None, password=None):
+  def authenticate(self, *args, **kwargs):
+    if py3:
```
Comment (romainr) [inline desktop/core/src/desktop/auth/backend.py:277]: py3 is not defined? Check the headers of other files for how to use `sys` module to check for the Python version

**Review by romainr — APPROVED** [review]
Not all the backends are converted but this is a good for now!

File: desktop/core/src/desktop/auth/backend.py [hunk 2]
```diff
@@ -273,7 +274,14 @@ class AllowFirstUserDjangoBackend(django.contrib.auth.backends.ModelBackend):
...
   """
-  def authenticate(self, username=None, email=None, password=None):
+  def authenticate(self, *args, **kwargs):
+    if sys.version_info[0] > 2:
+      request = args[0]
+
+    username = kwargs['username']
+    email = kwargs['email']
+    password = kwargs['password']
+
```
Comment (romainr) [inline desktop/core/src/desktop/auth/backend.py:284]: Actually, should remove line 288 and do

else:
   request = None

?

File: desktop/core/src/desktop/auth/backend.py [hunk 3]
```diff
@@ -318,7 +326,14 @@ class ImpersonationBackend(django.contrib.auth.backends.ModelBackend):
   Authenticate with a proxy user username/password but then login as another user.
   Does not support a multiple backends setup.
   """
-  def authenticate(self, username=None, password=None, login_as=None):
+  def authenticate(self, *args, **kwargs):
+    if sys.version_info[0] > 2:
+      request = args[0]
+
```
Comment (romainr) [inline desktop/core/src/desktop/auth/backend.py:332]: same

File: desktop/core/src/desktop/auth/backend.py [hunk 4]
```diff
@@ -273,11 +274,18 @@ class AllowFirstUserDjangoBackend(django.contrib.auth.backends.ModelBackend):
...
+  def authenticate(self, *args, **kwargs):
+    if sys.version_info[0] > 2:
+      request = args[0]
+    else:
+      request = None
+
+    username = kwargs['username']
+    password = kwargs['password']
+
+    if 'email' in kwargs:
```
Comment (agl29) [inline desktop/core/src/desktop/auth/backend.py:286]: some tests were failing so now I have to change like this. the error was no keyword 'email' in kwargs @romainr

File: desktop/core/src/desktop/auth/backend.py [hunk 5]
```diff
@@ -318,11 +326,19 @@ class ImpersonationBackend(django.contrib.auth.backends.ModelBackend):
...
-  def authenticate(self, username=None, password=None, login_as=None):
+  def authenticate(self, *args, **kwargs):
+    if sys.version_info[0] > 2:
+      request = args[0]
+    else:
+      request = None
+
+    username = kwargs['username']
+    password = kwargs['password']
+    login_as = kwargs['login_as']
```
Comment (agl29) [inline desktop/core/src/desktop/auth/backend.py:337]: and also here can we confirm that all three are keyword arguments? @romainr

Comment (romainr) [reply desktop/core/src/desktop/auth/backend.py:337]: The only way to know is by checking the call calling it if there are no tests for that

File: desktop/core/src/desktop/auth/backend.py (hunk 4 above)
Comment (romainr) [reply desktop/core/src/desktop/auth/backend.py:286]: :+1: This is one good day to port the previous `email is not None`

**Review by romainr — APPROVED** [review]

**Pull request merged**